File: 142-linked-list-cycle-ii/142-linked-list-cycle-ii.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        ListNode fastptr=head;
        ListNode slowptr=head;
        while(fastptr!=null && fastptr.next!=null){
            fastptr=fastptr.next.next;
            slowptr=slowptr.next;
            if(fastptr==slowptr){
                ListNode temp=head;
        
        
        while(temp!=slowptr){
           
        temp=temp.next;
        slowptr=slowptr.next;
            }
            return slowptr;  }
            
        } 
    return null;
    }}

File: 142-linked-list-cycle-ii/142-linked-list-cycle-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        fastptr=head
        slowptr=head
        if  head==None or head.next==None :
            return None
        while fastptr!=None and fastptr.next!=None:
            fastptr=fastptr.next.next
            slowptr=slowptr.next
            if slowptr==fastptr:
                temp=head
                while slowptr!=temp:
                    temp=temp.next
                    slowptr=slowptr.next
                return slowptr
        return None
                